cpp/src/io/utilities/datasource.cpp

```diff
@@ -135,6 +137,65 @@ class memory_mapped_source : public datasource {
   size_t map_offset_ = 0;
 };
 
+/**
+ * @brief Implementation class for reading from a file optimized for reading
+ * to device memory. Attempts to hide HtoD transfer latency with disk io by using
+ * a double-buffer.
+ *
+ **/
+class gpu_async_filereader : public datasource {
+  static constexpr uint32_t default_buffer_size = 16 << 20; // 16MB
+  struct pinned_filebuf {
+    uint8_t *host_bfr = nullptr;
+    cudaStream_t stream = 0;
+    cudaEvent_t completion_event = 0;
+    uint32_t position = 0;
+    bool event_pending = false;
+    ~pinned_filebuf() {
+      if (completion_event) {
+        cudaEventDestroy(completion_event);
+      }
+      if (host_bfr) {
+        cudaFreeHost(host_bfr);
+      }
+    }
+  };
+ public:
+  explicit gpu_async_filereader(const std::string& filepath, uint32_t bfrsize = default_buffer_size):
```

Comment:
I find it hard to read with no spaces between function definitions. Can you add a blank line after each function?

Reply:
ok